cardano-api/src/Cardano/Api/ProtocolParametersUpdate.hs

```diff
@@ -0,0 +1,674 @@
+        -- keys present in m2 and not m1: keep m2
+        -- I.e prefer m2 -- TODO: left off here. Remember plutus json jared sent you
+        mergeMap :: (ProtocolParametersUpdate -> Map AnyScriptLanguage a)
+                 -> Map AnyScriptLanguage a
+        mergeMap = undefined
+
+instance Monoid ProtocolParametersUpdate where
+    mempty =
+      ProtocolParametersUpdate {
+        protocolUpdateProtocolVersion     = Nothing
+      , protocolUpdateDecentralization    = Nothing
+      , protocolUpdateExtraPraosEntropy   = Nothing
+      , protocolUpdateMaxBlockHeaderSize  = Nothing
+      , protocolUpdateMaxBlockBodySize    = Nothing
+      , protocolUpdateMaxTxSize           = Nothing
+      , protocolUpdateTxFeeFixed          = Nothing
+      , protocolUpdateTxFeePerByte        = Nothing
+      , protocolUpdateMinUTxOValue        = Nothing
+      , protocolUpdateStakeAddressDeposit = Nothing
+      , protocolUpdateStakePoolDeposit    = Nothing
+      , protocolUpdateMinPoolCost         = Nothing
+      , protocolUpdatePoolRetireMaxEpoch  = Nothing
+      , protocolUpdateStakePoolTargetNum  = Nothing
+      , protocolUpdatePoolPledgeInfluence = Nothing
+      , protocolUpdateMonetaryExpansion   = Nothing
+      , protocolUpdateTreasuryCut         = Nothing
+      , protocolUpdateUTxOCostPerByte     = Nothing
+      , protocolUpdateCostModels          = mempty
+      , protocolUpdatePrices              = mempty
+      , protocolUpdateMaxTxExUnits        = Nothing
+      , protocolUpdateMaxBlockExUnits     = Nothing
+      , protocolUpdateParamMaxValSize     = Nothing
+      }
+
+-- ----------------------------------------------------------------------------
+-- Proposals embedded in transactions to update protocol parameters
+--
+
+data UpdateProposal =
+     UpdateProposal
+       !(Map (Hash GenesisKey) ProtocolParametersUpdate)
+       !EpochNo
+    deriving stock (Eq, Show)
+
+instance HasTypeProxy UpdateProposal where
+    data AsType UpdateProposal = AsUpdateProposal
+    proxyToAsType _ = AsUpdateProposal
+
+instance HasTextEnvelope UpdateProposal where
+    textEnvelopeType _ = "UpdateProposalShelley"
+
+--TODO: Jordan UpdateProposal needs to be parameterized by era or have access to the era
+instance SerialiseAsCBOR UpdateProposal where
+    serialiseToCBOR = CBOR.serializeEncoding' . toCBOR . toShelleyUpdate @StandardShelley
+    deserialiseFromCBOR _ bs =
+      fromShelleyUpdate @StandardShelley <$>
+        CBOR.decodeFull (LBS.fromStrict bs)
+
+
+makeShelleyUpdateProposal :: ProtocolParametersUpdate
+                          -> [Hash GenesisKey]
+                          -> EpochNo
+                          -> UpdateProposal
+makeShelleyUpdateProposal params genesisKeyHashes =
+    --TODO decide how to handle parameter validation
+    UpdateProposal (Map.fromList [ (kh, params) | kh <- genesisKeyHashes ])
+
+newtype CostModel = CostModel (Map.Map Text Integer)
```

Comment:
Oops, looks like we've duplicated these types from the `ProtocolParameters` module. A good reason to unsplit the modules (once we make the original less enormous again) :-)